### PR TITLE
Pause/Delete queue

### PR DESCRIPTION
**Pull request by pavan541cs**
- Created new endpoint to set status of queue
- Raise exception in case token is created for paused queue

This PR is to address #20

### CODE REVIEW

**Comment by pavan541cs** [comment]
@daltonfury42 Have a look when you have some time

File: complete/src/main/java/com/example/restservice/service/TokenService.java [hunk 1]
```diff
@@ -75,6 +76,9 @@ public TokenDetailResponse createToken(CreateTokenRequest createTokenRequest) {
             .findById(createTokenRequest.getQueueId())
             .map(
                 queue -> {
+                  if(queue.getStatus().equals(QueueStatus.PAUSED)) {
```
Comment (daltonfury42) [inline complete/src/main/java/com/example/restservice/service/TokenService.java:79]: let's also fail if the queue has been deleted

File: complete/src/main/java/com/example/restservice/constants/QueueStatus.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.example.restservice.constants;
+
+public enum QueueStatus {
+    ACTIVE,
+    PAUSED,
+    DELETED
```
Comment (daltonfury42) [inline complete/src/main/java/com/example/restservice/constants/QueueStatus.java:6]: Is deleting going to be a follow up PR?  Let's do it in this PR itself, I feel.

1. We should not allow a paused or active status to be set on a deleted queue.
2. If delete is doing to be a separate request as we had discussed, then we should not allow `PauseQueueRequest` to take deleted as a status in the body.

File: complete/src/main/java/com/example/restservice/service/QueueService.java [hunk 3]
```diff
@@ -41,6 +38,52 @@ public CreateQueueResponse createQueue(CreateQueueRequest createQueueRequest) {
     }
   }
 
+  @Transactional
+  public UpdateQueueStatusResponse pauseQueue(PauseQueueRequest pauseQueueRequest, String queueId) {
+    try {
```
Comment (daltonfury42) [inline complete/src/main/java/com/example/restservice/service/QueueService.java:43]: we should not allow it to set status to deleted through this endpoint

File: complete/src/main/java/com/example/restservice/service/QueueService.java [hunk 4]
```diff
@@ -41,6 +38,52 @@ public CreateQueueResponse createQueue(CreateQueueRequest createQueueRequest) {
...
+                if (!queue1.getOwnerId().equals(loggedInUserInfo.getUserId())) {
+                  throw new SQAccessDeniedException("You do not have access to this queue");
+                }
+                queue1.setStatus(pauseQueueRequest.getStatus());
+                return queueRepository.save(queue1);
+              }
+          ).get();
+      return new UpdateQueueStatusResponse(queue.getQueueId(), queue.getQueueName(), queue.getStatus());
+    } catch (DataIntegrityViolationException de) {
+      throw SQInvalidRequestException.queueNameNotUniqueException();
```
Comment (daltonfury42) [inline complete/src/main/java/com/example/restservice/service/QueueService.java:58]: this doesn't make sense

File: complete/src/main/java/com/example/restservice/service/QueueService.java [hunk 5]
```diff
@@ -41,6 +38,52 @@ public CreateQueueResponse createQueue(CreateQueueRequest createQueueRequest) {
...
+                        if (!queue1.getOwnerId().equals(loggedInUserInfo.getUserId())) {
+                            throw new SQAccessDeniedException("You do not have access to this queue");
+                        }
+                        queue1.setStatus(QueueStatus.DELETED);
+                        return queueRepository.save(queue1);
+                    }
+                ).get();
+            return new UpdateQueueStatusResponse(queue.getQueueId(), queue.getQueueName(), queue.getStatus());
+        } catch (DataIntegrityViolationException de) {
+            throw SQInvalidRequestException.queueNameNotUniqueException();
```
Comment (daltonfury42) [inline complete/src/main/java/com/example/restservice/service/QueueService.java:81]: remove

File: complete/src/main/java/com/example/restservice/service/QueueService.java [hunk 6]
```diff
@@ -41,6 +38,52 @@ public CreateQueueResponse createQueue(CreateQueueRequest createQueueRequest) {
...
+        try {
+            var queue =
+                queueRepository
+                .findById(queueId)
+                .map(
+                    queue1 -> {
+                        if (!queue1.getOwnerId().equals(loggedInUserInfo.getUserId())) {
+                            throw new SQAccessDeniedException("You do not have access to this queue");
+                        }
+                        queue1.setStatus(QueueStatus.DELETED);
```
Comment (daltonfury42) [inline complete/src/main/java/com/example/restservice/service/QueueService.java:75]: If queue was already deleted once, should we just throw? I think we should

File: complete/src/main/java/com/example/restservice/service/QueueService.java [hunk 7]
```diff
@@ -41,6 +38,52 @@ public CreateQueueResponse createQueue(CreateQueueRequest createQueueRequest) {
...
+    try {
+      var queue =
+          queueRepository
+          .findById(queueId)
+          .map(
+              queue1 -> {
+                if (!queue1.getOwnerId().equals(loggedInUserInfo.getUserId())) {
+                  throw new SQAccessDeniedException("You do not have access to this queue");
+                }
+                queue1.setStatus(pauseQueueRequest.getStatus());
```
Comment (daltonfury42) [inline complete/src/main/java/com/example/restservice/service/QueueService.java:52]: lets throw here if queue was deleted.

File: complete/src/main/java/com/example/restservice/service/TokenService.java (hunk 1 above)
Comment (daltonfury42) [reply complete/src/main/java/com/example/restservice/service/TokenService.java:79]: open to discussion on this one, though. If a queue was paused once, should we allow it to be paused again.? I think not. Coz if it's called again, there is a bug in the UI?

**Comment by pavan541cs** [comment]
@daltonfury42 Addressed the comments

File: complete/src/main/java/com/example/restservice/exceptions/SQInvalidRequestException.java [hunk 8]
```diff
@@ -2,23 +2,31 @@
...
+    TOKEN_DELETED,
+    QUEUE_PAUSED,
+    QUEUE_DELETED,
+    QUEUE_DELETE_NOT_ALLOWED
   }
 
-  private final ImmutableMap<ReasonCode, String> message =
-      ImmutableMap.of(
+  private final Map<ReasonCode, String> message =
+      Map.of(
```
Comment (daltonfury42) [inline complete/src/main/java/com/example/restservice/exceptions/SQInvalidRequestException.java:21]: just curious, why not ImmutableMap?

Comment (pavan541cs) [reply complete/src/main/java/com/example/restservice/exceptions/SQInvalidRequestException.java:21]: Immutable map has a restriction of having max of five keys

**Comment by daltonfury42** [comment]
I'm playing with this tool called Flyway. It will do automatic schema migrations... I want to merge this change after setting up flyway, and then I can check if the flyway setup worked.

I'm mostly merge by end of week.

**Comment by daltonfury42** [comment]
@pavan541cs thanks for the PR!!

**Pull request merged**